### ISSUE TITLE
Custom Dataset distributed training

### ISSUE THREAD

**Issue by ParadiseN1** [chengtan9907/OpenSTL#22]
Hi, first of all thank you for a great work on SimVP and really nice repository.

Im currently want to to train a SimVP model on ocean data to forecast ocean transparency. I found out that on T4 GPU training is realy slow on my data, one iteration takes approx. 0.7 seconds, so i decided to train it on multiple GPU, but faced a problem that every process is creating separate entity of Dataset and end up with SIGKILL, caused by out of RAM while loading separate Datasets for every GPU. 

I have no experience with distributed trainings so, maybe i messed something up. Maybe you have some ideas on how to solve this issue, it would be very helpful!

**Comment by ParadiseN1** [comment]
This is how my Dataset file looks like:
`from torch.utils.data import Dataset
import os.path as osp
import numpy as np
import xarray as xr
import torch
from PIL import Image
import random
import time

d2r = np.pi / 180

AVG_NAN_COUNT = 33659

def latlon2xyz(lat, lon):
    if type(lat) == torch.Tensor:
        x = -torch.cos(lat)*torch.cos(lon)
        y = -torch.cos(lat)*torch.sin(lon)
        z = torch.sin(lat)

    if type(lat) == np.ndarray:
        x = -np.cos(lat)*np.cos(lon)
        y = -np.cos(lat)*np.sin(lon)
        z = np.sin(lat)
    return x, y, z


def xyz2latlon(x, y, z):
    if type(x) == torch.Tensor:
        lat = torch.arcsin(z)
        lon = torch.atan2(-y, -x)

    if type(x) == np.ndarray:
        lat = np.arcsin(z)
        lon = np.arctan2(-y, -x)
    return lat, lon

def drop_last_column_if_odd(data):
    lat_size = data.lat.size
    lon_size = data.lon.size

    lat_slice = slice(None, -1) if lat_size % 2 != 0 else slice(None)
    lon_slice = slice(None, -1) if lon_size % 2 != 0 else slice(None)

    return data.isel(lat=lat_slice, lon=lon_slice)

class ZsdDataset(Dataset):
    def __init__(self, data_dir, training_time, seq_len=10, out_seq_len=1, surface_mask=True, mean=None, std=None):
        super().__init__()
        self.y_start = training_time[0]
        self.y_end = training_time[1]
        self.seq_len = 10
        self.out_seq_len = out_seq_len
        self.time = None
        self.threshold = AVG_NAN_COUNT * 1.1 # empirical value, when all datapoints with NaNs due to sun illumination are above this threshold
        self.patch_h = 32
        self.patch_w = 64
        self.mean = mean
        self.std = std

        print(data_dir)
        try:
            dataset = xr.open_mfdataset(
                data_dir+'/*.nc', combine='by_coords')
        except AttributeError:
            assert False and 'Please install the latest xarray, e.g.,' \
                                'pip install  git+https://github.com/pydata/xarray/@v2022.03.0'
        dataset = drop_last_column_if_odd(dataset)
        dataset = dataset.sel(time=slice(self.y_start, self.y_end))
        
        self.num_patches_per_image = len(dataset.lat.values)//self.patch_h * len(dataset.lon.values)//self.patch_w
        
        ## Filter datapoints with NaNs
        missing_values_count = dataset['ZSD'].isnull().sum(dim=['lon', 'lat'])
        missing_values_count1 = dataset['CHL'].isnull().sum(dim=['lon', 'lat'])
        timestamps_below_threshold = (missing_values_count < self.threshold)
        print(f'Filtered {timestamps_below_threshold.values.sum()}/{len(timestamps_below_threshold.values)}')
        dataset = dataset.where(timestamps_below_threshold, drop=True)
        


        self.zsd_data = dataset.get('ZSD').values[:, np.newaxis, :, :]
        self.chl_data = dataset.get('CHL').values[:, np.newaxis, :, :]
        
        self.zsd_data = np.concatenate([self.zsd_data, self.chl_data], axis=1)
        if surface_mask:
            self.surface_mask = np.array(Image.open(data_dir+"/surface_mask.png"))[...,0] / 255.
            self.zsd_data = self.zsd_data*self.surface_mask[:self.zsd_data.shape[-2], :self.zsd_data.shape[-1]]
            

        if self.mean is None:
            self.mean = self.zsd_data[self.zsd_data>0].mean()
            self.std = self.zsd_data[self.zsd_data>0].std()
        self.zsd_data = (self.zsd_data-self.mean)/self.std
        self.zsd_data = np.nan_to_num(self.zsd_data, nan=-777)

        self.valid_idx = np.array(
            range(0, (self.zsd_data.shape[0]- out_seq_len - self.seq_len) * self.num_patches_per_image))
    def __len__(self):
        return self.valid_idx.shape[0]

    def __getitem__(self, index):
        # start_time = time.time()
        idx = (self.valid_idx[index] // self.num_patches_per_image)
        # get random patch from image
        while True:
            lon_i = int(random.uniform(0, self.zsd_data.shape[-2]-self.patch_h))
            lat_i = int(random.uniform(0, self.zsd_data.shape[-1]-self.patch_w))
            y_data = torch.tensor(self.zsd_data[idx+self.seq_len:idx+self.seq_len+self.out_seq_len, 0, lon_i:lon_i+self.patch_h, lat_i:lat_i+self.patch_w]).to(torch.float32)[None]
            if y_data.eq(-777).all():
                continue
            break
        x_data = torch.tensor(self.zsd_data[idx:idx+self.seq_len, :, lon_i:lon_i+self.patch_h, lat_i:lat_i+self.patch_w]).to(torch.float32)
        # elapsed_time = time.time() - start_time
        # print(f"Load batch took {elapsed_time} seconds to run.")
        return x_data, y_data

from openstl.datasets.utils import create_loader

def load_data(batch_size, val_batch_size, data_root, num_workers=4, data_name='mnist',
              pre_seq_length=10, aft_seq_length=10, in_shape=[10, 1, 64, 64],
              distributed=False, use_augment=False, use_prefetcher=False, train_time=['2010', '2021'],
              val_time=['2020', '2021'],
              test_time=['2021', '2023'], **kwargs):
    train_set = ZsdDataset(data_dir=data_root, training_time=train_time)
    # validation_set = ZsdDataset(data_root, val_time, mean=train_set.mean, std=train_set.std)
    test_set = ZsdDataset(data_root, test_time, mean=train_set.mean, std=train_set.std)


    dataloader_train = create_loader(train_set,
                                     batch_size=batch_size,
                                     shuffle=True, is_training=True,
                                     pin_memory=True, drop_last=True,
                                     num_workers=num_workers,
                                     distributed=distributed, use_prefetcher=use_prefetcher)
    dataloader_vali = create_loader(test_set,
                                    batch_size=val_batch_size,
                                    shuffle=False, is_training=False,
                                    pin_memory=True, drop_last=False,
                                    num_workers=num_workers,
                                    distributed=distributed, use_prefetcher=use_prefetcher)
    dataloader_test = create_loader(test_set,
                                    batch_size=val_batch_size,
                                    shuffle=False, is_training=False,
                                    pin_memory=True, drop_last=False,
                                    num_workers=num_workers,
                                    distributed=distributed, use_prefetcher=use_prefetcher)
    return dataloader_train, dataloader_vali, dataloader_test`

**Comment by BraveDistribution** [comment]
same issue with the pytorch lightning.

**Comment by Lupin1998** [comment]
> Hi, first of all thank you for a great work on SimVP and really nice repository.
> 
> Im currently want to to train a SimVP model on ocean data to forecast ocean transparency. I found out that on T4 GPU training is realy slow on my data, one iteration takes approx. 0.7 seconds, so i decided to train it on multiple GPU, but faced a problem that every process is creating separate entity of Dataset and end up with SIGKILL, caused by out of RAM while loading separate Datasets for every GPU.
> 
> I have no experience with distributed trainings so, maybe i messed something up. Maybe you have some ideas on how to solve this issue, it would be very helpful!

Hi, @ParadiseN1, very sorry for the late reply. Actually, we have also met some out of RAM memory problems when we train on a GPU node with small RAM (e.g., 128G). We have fixed an out-of-memory issue in the validation process in the latest commit, but I cannot provide a helpful solution to solve the out-of-memory issue when loading the whole dataset. Maybe you can try this way: you can load the index file at the dataset initialization instead of loading the whole data, and then read the data indicated by the index in `__getitem__(self, index)`. In this solution, the mean and std should be pre-computed and loaded at the initialization.

**Comment by chengtan9907** [comment]
I will close this issue if there is no more question. Please feel free to open a new issue when you have new questions.
